doc/api/n-api.md

```diff
@@ -3088,6 +3088,52 @@ The native string is copied.
 The JavaScript `string` type is described in
 [Section 6.1.4][] of the ECMAScript Language Specification.
 
+### Functions to create optimized property keys
+
+Many JavaScript engines including V8 use internalized strings as keys
+to set and get property values. They typically use a hash table to create
+and lookup such strings. While it adds some cost per key creation, it improves
+the performance after that by enabling comparison of string pointers instead
+of the whole strings.
+
+If a new JavaScript string is intended to be used as a property key, then for
+some JavaScript engines it will be more efficient to use the function in this
+section. Otherwise, use the `napi_create_string_utf8` or
+`node_api_create_external_string_utf8` series functions as there may be
```

Comment:
Why? If the engine doesn't provide a benefit when calling this function for any given implementation of Node-API, then, for that implementation, these functions should fall back to just calling `napi_create_string_*`, which should not incur no more than the burden of an additional indirection.

Reply:
There is a slight overhead for creating the internalized string because besides allocating the space for the string we should search for a string with the same content in the hash table. Depending on the string size, memory allocator, and number of strings in the hash table, the internalized strings could be more expensive or less expensive than normal string.
I guess for the small strings which can benefit from the small memory allocators, the lookup in the hash table is usually more expensive.

Reply:
In our meeting today we decided to update the documentation for all these APIs after we land this PR.